compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp

```diff
@@ -174,6 +176,225 @@ void transform_dialect::ApplyFoldFillIntoPadPatternsOp::populatePatterns(
+        }
+        destSlice = maybeSlice;
+      }
+      // Verify they operate on equivalent subsets, ensuring the slices are
+      // hoistable. It is still possible to hoist the loop if this is not true,
+      // however in such cases we likely formed the loops in the wrong order.
+      if (!cast<SubsetOpInterface>(*destSlice)
+               .operatesOnEquivalentSubset(
+                   cast<SubsetOpInterface>(*parallelInsert),
+                   [](Value v1, Value v2) { return v1 == v2; })) {
+        return failure();
+      }
+      terminators[destBbArg.getArgNumber() - numInductionVars] = parallelInsert;
+      pairedSlices[destBbArg.getArgNumber() - numInductionVars] = destSlice;
+
+      // Collect all of the offset/size/stride operands for both slices and
+      // compute a backwards slice of the program from them. Fail if any of
+      // them depend on the serial loop iterator.
+      llvm::SmallDenseSet<Value> sliceOperands;
+      sliceOperands.insert(
+          parallelInsert.getOperands().begin() +
+              parallelInsert.getOffsetSizeAndStrideStartOperandIndex(),
+          parallelInsert.getOperands().end());
+      sliceOperands.insert(
+          destSlice.getOperands().begin() +
+              destSlice.getOffsetSizeAndStrideStartOperandIndex(),
+          destSlice.getOperands().end());
+      for (Value operand : sliceOperands) {
+        if (auto bbArg = dyn_cast<BlockArgument>(operand)) {
+          if (bbArg.getOwner()->getParentOp() == loop) {
+            return rewriter.notifyMatchFailure(
+                loop, "Slice operand producers depend on loop");
+          }
+        }
+        SetVector<Operation *> tmpBackwardSlice;
+        getBackwardSlice(operand, &tmpBackwardSlice, backwardOptions);
+        slice.set_union(tmpBackwardSlice);
+      }
+    }
+
+    // An operation is safe to hoist if it is speculatable and none of its
+    // operands depend on the serial loop.
+    auto isSafeToHoist = [&](Operation *op) {
+      return op->getBlock() == forallOp.getBody() && isMemoryEffectFree(op) &&
+             isSpeculatable(op) &&
+             llvm::none_of(op->getOperands(), [&](Value operand) {
+               if (auto blockArg = dyn_cast<BlockArgument>(operand)) {
+                 return blockArg.getOwner() == loop.getBody();
+               }
+               return false;
+             });
+    };
+
+    if (!llvm::all_of(slice, isSafeToHoist)) {
+      return rewriter.notifyMatchFailure(
+          loop, "Slice operand producers not safe to hoist out of loop");
+    }
+
+    // Sort the backwards slice of the producers for the insertion/extraction
+    // indices by the block order in the scf.forall body. This ensures that we
+    // hoist operations in the same order they started. Any topological ordering
+    // would work too because the operations are speculatable.
+    SmallVector<Operation *> sliceVec(slice.getArrayRef());
+    std::sort(sliceVec.begin(), sliceVec.end(),
```

Comment:
Just use the`mlir::topologicalSort` routine?

Reply:
oh cool, I didn't know this existed